src/pudl/extract/eia930.py

```diff
@@ -1,97 +1,145 @@
+        if (year > 2018) or (year == 2018 and half == 2)
+    ]
+    return extract_page(
+        datastore=context.resources.datastore,
+        page="subregion",
+        half_years=valid_half_years,
+    )
+
+
+def extract_page(datastore: Datastore, page: str, half_years: list[str]) -> Path:
+    """Pull data for a page across many half-years into a Parquet file.
+
+    This involves reading each half-year, of course, but also concatenating them
+    together and expanding the schema to fit all the columns we see.
+
+    If we were to return the `con.query()` and use an IOManager to manage
+    the Parquet IO, we would have to manage the DuckDB connection lifetime
+    so we don't try to write out from a closed DuckDB connection. So, we just
+    write out directly in this asset and return a Path that we can pass to
+    ``pd.read_parquet``, ``pl.scan_parquet``, or any other Parquet reading
+    strategy.
+
+    Args:
+        datastore: the Datastore we use to actually access the raw data.
+        page: the name of the page we're extracting.
+        half_years: the set of half-year segments we're extracting.
+
+    Returns:
+        The path to the resulting Parquet file.
+    """
+    con = duckdb.connect()
+    individual_views = [
+        extract_half_year_page(
+            con,
+            datastore=datastore,
+            half_year=half_year,
+            page=page,
+        )
+        for half_year in half_years
+    ]
+    union_query = " UNION ALL BY NAME ".join(
+        f"SELECT * FROM {view_name}"  # noqa: S608 (we trust this view name)
+        for view_name in individual_views
+    )
+    output_path = PudlPaths().parquet_path(f"raw_eia930__{page}.parquet")
+    con.query(union_query).to_parquet(str(output_path))
+    return output_path.with_suffix("")
+
+
+def extract_half_year_page(
+    con: duckdb.DuckDBPyConnection,
+    datastore: Datastore,
+    half_year: str,
+    page: str,
+) -> str:
+    """Extract data from a single CSV.
+
+    Reads into DuckDB for speed and memory use. To avoid reading the whole CSV
+    into memory, we're extracting directly to a temporary directory.
+
+    Args:
+        con: DuckDB connection.
+        datastore: the Datastore we use to actually access the input data.
+        half_year: the half-year we're reading in.
+        page: the name of the page we're reading.
+
+    Returns:
+        view_name: name of DuckDB view that represents the read & renamed CSV.
+    """
+    dataset_name = "eia930"
+    filename = f"eia930-{half_year}-{page}.csv"
+    metadata = GenericMetadata(dataset_name)
+    column_map = metadata.get_column_map(page=page, half_year=half_year)
+
+    with datastore.get_zipfile_resource(dataset_name, half_year=half_year) as zf:
+        unzipped = zf.extract(filename, datastore.temporary_extraction_dir.name)
+    csv_rel = con.read_csv(unzipped)
+
+    def clean_name(name):
+        return re.sub(r"\W+", "_", name.lower()).strip("_")
```

Comment:
I found https://github.com/catalyst-cooperative/pudl/blob/39b4c71925ad2febb8ffb25181edaccd168683b2/src/pudl/helpers.py#L1100

is that useful here?

Reply:
It's specialized for pandas dataframes unfortunately. `DuckDBPyRelation.columns` is just a normal dict and doesn't have the `.str` methods on it.